File: objectModel/Python/tests/utilities/projection_test_utils.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See License.txt in the project root for license information.

import os
from typing import List, Optional

from cdm.enums import CdmLogCode, CdmObjectType, CdmStatusLevel
from cdm.utilities import AttributeResolutionDirectiveSet, ResolveOptions
from tests.cdm.projection.attribute_context_util import AttributeContextUtil
from tests.common import TestHelper


shortened_directives = {
    'normalized': 'norm',
    'referenceOnly': 'refOnly',
    'structured': 'struc',
    'virtual': 'virt',
}

class ProjectionTestUtils:
    """
    Common utility methods for projection tests
    If you want to update the expected output txt files for all the tests that are ran,
    please set the parameter update_expected_output true in the method validate_attribute_context()
    """

    # Path to foundations
    foundation_json_path = 'cdm:/foundations.cdm.json'

    # The log codes that are allowed to be logged without failing the test
    allowed_logs = set([
        CdmLogCode.WARN_DEPRECATED_RESOLUTION_GUIDANCE.name
    ])

    @staticmethod
    async def get_resolved_entity(corpus: 'CdmCorpusDefinition', input_entity: 'CdmEntityDefinition', directives: List[str]) -> 'CdmEntityDefinition':
        """Resolves an entity"""
        ro_hash_set = set(directives)

        resolved_entity_name = 'Resolved_{}'.format(input_entity.entity_name)

        res_opt = ResolveOptions(input_entity.in_document, directives=AttributeResolutionDirectiveSet(ro_hash_set))

        resolved_folder = corpus.storage.fetch_root_folder('output')
        resolved_entity = await input_entity.create_resolved_entity_async(resolved_entity_name, res_opt, resolved_folder)  # type: CdmEntityDefinition

        return resolved_entity

    @staticmethod
    def get_resolution_option_name_suffix(directives: List[str], expected_output_path = None, entity_name = None) -> str:
        """Returns a suffix that contains the file name and resolution option used"""
        file_name_prefix = ''

        for directive in directives:
            shortened_directive = shortened_directives.get(directive)
            if not shortened_directive:
                raise Exception('Using unsupported directive')

            file_name_prefix = '{}_{}'.format(file_name_prefix, shortened_directive)

        file_exists = os.path.exists(os.path.join(expected_output_path, 'AttrCtx_{}{}.txt'.format(entity_name, file_name_prefix))) \
            if expected_output_path and entity_name else True

        if not file_name_prefix or not file_exists:
            file_name_prefix = '_default'

        return file_name_prefix

    @staticmethod
    async def load_entity_for_resolution_option_and_save(test: 'TestCase', corpus: 'CdmCorpusDefinition', test_name: str, tests_subpath: str, entity_name: str, \
                                                         directives: List[str], update_expected_output: bool = False) -> 'CdmEntityDefinition':
        """Loads an entity, resolves it, and then validates the generated attribute contexts"""
        expected_output_path = TestHelper.get_expected_output_folder_path(tests_subpath, test_name)

        entity = await corpus.fetch_object_async('local:/{0}.cdm.json/{0}'.format(entity_name))
        test.assertIsNotNone(entity)
        resolved_entity = await ProjectionTestUtils.get_resolved_entity(corpus, entity, directives)
        test.assertIsNotNone(resolved_entity)

        await ProjectionTestUtils.validate_attribute_context(test, directives, expected_output_path, entity_name, resolved_entity, update_expected_output)

        return resolved_entity

    @staticmethod
    def get_local_corpus(tests_subpath: str, test_name: str) -> 'CdmCorpusDefinition':
        """Creates a corpus"""
        corpus = TestHelper.get_local_corpus(tests_subpath, test_name)

        def callback(level: CdmStatusLevel, message: str):
            last_event = corpus.ctx.events[-1]
            if not last_event.get('code') or last_event['code'] not in ProjectionTestUtils.allowed_logs:
                raise Exception(message)
        corpus.set_event_callback(callback, CdmStatusLevel.WARNING)

        return corpus

    @staticmethod
    def create_entity(corpus: 'CdmCorpusDefinition', local_root: 'CdmFolderDefinition') -> 'CdmEntityDefinition':
        """Creates an entity"""
        entity_name = 'TestEntity'
        entity = corpus.make_object(CdmObjectType.ENTITY_DEF, entity_name)

        entity_doc = corpus.make_object(CdmObjectType.DOCUMENT_DEF, '{}.cdm.json'.format(entity_name), False)
        entity_doc.imports.append(ProjectionTestUtils.foundation_json_path)
        entity_doc.definitions.append(entity)
        local_root.documents.append(entity_doc, entity_doc.name)

        return entity

    @staticmethod
    def create_source_entity(corpus: 'CdmCorpusDefinition', local_root: 'CdmFolderDefinition') -> 'CdmEntityDefinition':
        """Creates a source entity for a projection"""
        entity_name = 'SourceEntity'
        entity = corpus.make_object(CdmObjectType.ENTITY_DEF, entity_name)

        attribute_name1 = 'id'
        attribute1 = corpus.make_object(CdmObjectType.TYPE_ATTRIBUTE_DEF, attribute_name1)
        attribute1.date_type = corpus.make_ref(CdmObjectType.DATA_TYPE_REF, 'string', True)
        entity.attributes.append(attribute1)

        attributeName2 = 'name'
        attribute2 = corpus.make_object(CdmObjectType.TYPE_ATTRIBUTE_DEF, attributeName2)
        attribute2.date_type = corpus.make_ref(CdmObjectType.DATA_TYPE_REF, 'string', True)
        entity.attributes.append(attribute2)

        attributeName3 = 'value'
        attribute3 = corpus.make_object(CdmObjectType.TYPE_ATTRIBUTE_DEF, attributeName3)
        attribute3.date_type = corpus.make_ref(CdmObjectType.DATA_TYPE_REF, 'integer', True)
        entity.attributes.append(attribute3)

        attributeName4 = 'date'
        attribute4 = corpus.make_object(CdmObjectType.TYPE_ATTRIBUTE_DEF, attributeName4)
        attribute4.date_type = corpus.make_ref(CdmObjectType.DATA_TYPE_REF, 'date', True)
        entity.attributes.append(attribute4)

        entity_doc = corpus.make_object(CdmObjectType.DOCUMENT_DEF, '{}.cdm.json'.format(entity_name), False)
        entity_doc.imports.append(ProjectionTestUtils.foundation_json_path)
        entity_doc.definitions.append(entity)
        local_root.documents.append(entity_doc, entity_doc.name)

        return entity

    @staticmethod
    def create_projection(corpus: 'CdmCorpusDefinition', local_root: 'CdmFolderDefinition') -> 'CdmProjection':
        """Creates a projection"""
        # Create an entity reference to use as the source of the projection
        projection_source = corpus.make_object(CdmObjectType.ENTITY_REF, None)
        projection_source.explicit_reference = ProjectionTestUtils.create_source_entity(corpus, local_root)

        # Create the projection
        projection = corpus.make_object(CdmObjectType.PROJECTION_DEF)
        projection.source = projection_source

        return projection

    @staticmethod
    def validate_expansion_info_trait(test: 'TestCase', attribute: 'CdmTypeAttributeDefinition', expected_attr_name: str, ordinal: int, expansion_name: str, member_attribute: str):
        test.assertEqual(expected_attr_name, attribute.name)
        trait = attribute.applied_traits.item('has.expansionInfo.list')
        test.assertIsNotNone(trait)
        test.assertEqual(expansion_name, trait.arguments.fetch_value('expansionName'))
        test.assertEqual(str(ordinal), trait.arguments.fetch_value('ordinal'))
        test.assertEqual(member_attribute, trait.arguments.fetch_value('memberAttribute'))

    @staticmethod
    def validate_attribute_group(test: 'TestCase', attributes: 'CdmCollection[CdmAttributeItem]', attribute_group_name: str, \
                                 attributes_size: int = 1, index: int = 0) -> 'CdmAttributeGroupDefinition':
        """Validates the creation of an attribute group and return its definition
            @param attributes = The collection of attributes.
            @param attribute_group_name = The attribute group name.
            @param attributes_size = The expected size of the attributes collection."""
        test.assertEqual(attributes_size, len(attributes))
        test.assertEqual(CdmObjectType.ATTRIBUTE_GROUP_REF, attributes[index].object_type)
        att_group_reference = attributes[index]  # type: CdmAttributeGroupReference
        test.assertIsNotNone(att_group_reference.explicit_reference)

        att_group_definition = att_group_reference.explicit_reference  # type: CdmAttributeGroupDefinition
        test.assertEqual(attribute_group_name, att_group_definition.attribute_group_name)

        return att_group_definition

    @staticmethod
    async def validate_attribute_context(test: 'TestCase', directives: List[str], expected_output_path: str, entity_name: str, resolved_entity: 'CdmEntityDefinition', update_expected_output: Optional[bool] = False):
        """
        Validates if the attribute context of the resolved entity matches the expected output.
        If update_expected_output is true, will update the expected output txt files for all the tests that are ran.
        """
        if not resolved_entity.attribute_context:
            raise Exception('ValidateAttributeContext called with not resolved entity.')

        file_name_prefix = 'AttrCtx_' + entity_name
        file_name_suffix = ProjectionTestUtils.get_resolution_option_name_suffix(directives)

        # Get actual text
        attr_ctx_util = AttributeContextUtil()
        actual_text = attr_ctx_util.get_attribute_context_strings(resolved_entity)

        if update_expected_output:
            expected_string_file_path = os.path.join(expected_output_path, file_name_prefix + file_name_suffix + '.txt')

            if len(directives) > 0:
                default_file_name_suffix = ProjectionTestUtils.get_resolution_option_name_suffix([])
                default_string_file_path = os.path.join(expected_output_path, file_name_prefix + default_file_name_suffix + '.txt')
                if os.path.exists(default_string_file_path):
                    with open(default_string_file_path) as default_file:
                        default_text = default_file.read().replace('\r\n', '\n')
                else:
                    default_text = None

                if actual_text == default_text:
                    if os.path.exists(expected_string_file_path):
                        os.remove(expected_string_file_path)
                else:
                    with open(expected_string_file_path, 'w') as expected_file:
                        expected_file.write(actual_text)
            else:
                with open(expected_string_file_path, 'w') as expected_file:
                    expected_file.write(actual_text)
        else:
            # Actual
            actual_string_file_path = os.path.join(expected_output_path, '..', TestHelper.get_test_actual_output_folder_name(), file_name_prefix + file_name_suffix + '.txt')

            # Save Actual AttrCtx_*.txt and Resolved_*.cdm.json
            with open(actual_string_file_path, 'w') as expected_file:
                expected_file.write(actual_text)
            await resolved_entity.in_document.save_as_async(resolved_entity.entity_name + file_name_suffix + '.cdm.json', save_referenced=False)

            # Expected
            expected_file_name_suffix = ProjectionTestUtils.get_resolution_option_name_suffix(directives, expected_output_path, entity_name)
            expected_string_file_path = os.path.join(expected_output_path, file_name_prefix + expected_file_name_suffix + '.txt')
            with open(expected_string_file_path) as expected_file:
                expected_text = expected_file.read()

            # Test if Actual is Equal to Expected
            test.assertEqual(expected_text.replace('\r\n', '\n'), actual_text.replace('\r\n', '\n'))

File: objectModel/Java/objectmodel/src/test/java/com/microsoft/commondatamodel/objectmodel/utilities/ProjectionTestUtils.java
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

package com.microsoft.commondatamodel.objectmodel.utilities;

import com.microsoft.commondatamodel.objectmodel.TestHelper;
import com.microsoft.commondatamodel.objectmodel.cdm.*;
import com.microsoft.commondatamodel.objectmodel.cdm.projection.AttributeContextUtil;
import com.microsoft.commondatamodel.objectmodel.cdm.projections.CdmProjection;
import com.microsoft.commondatamodel.objectmodel.enums.CdmLogCode;
import com.microsoft.commondatamodel.objectmodel.enums.CdmObjectType;
import com.microsoft.commondatamodel.objectmodel.enums.CdmStatusLevel;
import com.microsoft.commondatamodel.objectmodel.utilities.logger.EventList;

import org.testng.Assert;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

/**
 * Common utility methods for projection tests
 * If you want to update the expected output txt files for all the tests that are ran,
 * please set the parameter updateExpectedOutput true in the method
 * @see ProjectionTestUtils#validateAttributeContext(List, String, String, CdmEntityDefinition, boolean)
 */
public class ProjectionTestUtils {
    /**
     * Path to foundations
     */
    private static final String foundationJsonPath = "cdm:/foundations.cdm.json";

    /**
     * Resolves an entity
     * @param corpus The corpus
     * @param inputEntity The entity to resolve
     * @param directives The set of directives used for resolution
     */
    public static CompletableFuture<CdmEntityDefinition> getResolvedEntity(
        CdmCorpusDefinition corpus,
        CdmEntityDefinition inputEntity,
        List<String> directives
    ) {
        return CompletableFuture.supplyAsync(() -> {
            HashSet<String> roHashSet = new HashSet<>(directives);

            String resolvedEntityName = "Resolved_" + inputEntity.getEntityName();

            ResolveOptions resOpt = new ResolveOptions(inputEntity.getInDocument());
            resOpt.setDirectives(new AttributeResolutionDirectiveSet(roHashSet));

            CdmFolderDefinition resolvedFolder = corpus.getStorage().fetchRootFolder("output");
            CdmEntityDefinition resolvedEntity = inputEntity.createResolvedEntityAsync(resolvedEntityName, resOpt, resolvedFolder).join();

            return resolvedEntity;
        });
    }

    /**
     * Returns a suffix that contains the file name and resolution option used
     * @param directives The set of directives used for resolution
     */
    public static String getResolutionOptionNameSuffix(List<String> directives) {
        String fileNamePrefix = "";

        for (String directive : directives) {
            String shortenedDirective;
            switch (directive) {
                case "normalized":
                    shortenedDirective = "norm";
                    break;
                case "referenceOnly":
                    shortenedDirective = "refOnly";
                    break;
                case "structured":
                    shortenedDirective = "struc";
                    break;
                case "virtual":
                    shortenedDirective = "virt";
                    break;
                default:
                    Assert.fail("Using unsupported directive");
                    return null;
            }
            fileNamePrefix = fileNamePrefix + "_" + shortenedDirective;
        }

        if (StringUtils.isNullOrTrimEmpty(fileNamePrefix)) {
            fileNamePrefix = "_default";
        }

        return fileNamePrefix;
    }

    /**
     * Loads an entity, resolves it, and then validates the generated attribute contexts
     */
    public static CompletableFuture<CdmEntityDefinition> loadEntityForResolutionOptionAndSave(final CdmCorpusDefinition corpus, final String testName, final String testsSubpath, final String entityName, List<String> directives) {
        String expectedOutputPath = null;
        try {
            expectedOutputPath = TestHelper.getExpectedOutputFolderPath(testsSubpath, testName);
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        CdmEntityDefinition entity = (CdmEntityDefinition) corpus.fetchObjectAsync("local:/" + entityName + ".cdm.json/" + entityName).join();
        Assert.assertNotNull(entity);
        CdmEntityDefinition resolvedEntity = ProjectionTestUtils.getResolvedEntity(corpus, entity, directives).join();
        Assert.assertNotNull(resolvedEntity);

        validateAttributeContext(directives, expectedOutputPath, entityName, resolvedEntity);

        return CompletableFuture.completedFuture(resolvedEntity);
    }

    public static CdmAttributeGroupDefinition validateAttributeGroup(CdmCollection<CdmAttributeItem> attributes, String attributeGroupName) {
        return validateAttributeGroup(attributes, attributeGroupName, 1, 0);
    }

    public static CdmAttributeGroupDefinition validateAttributeGroup(CdmCollection<CdmAttributeItem> attributes, String attributeGroupName, int attributesSize) {
        return validateAttributeGroup(attributes, attributeGroupName, attributesSize, 0);
    }

    /**
     * Validates the creation of an attribute group and return its definition
     * @param attributes The collection of attributes
     * @param attributeGroupName The attribute group name
     * @param attributesSize The expected size of the attributes collection
     */
    public static CdmAttributeGroupDefinition validateAttributeGroup(CdmCollection<CdmAttributeItem> attributes, String attributeGroupName, int attributesSize, int index) {
        Assert.assertEquals(attributesSize, attributes.getCount());
        Assert.assertEquals(CdmObjectType.AttributeGroupRef, attributes.get(index).getObjectType());
        CdmAttributeGroupReference attGroupReference = (CdmAttributeGroupReference) attributes.get(index);
        Assert.assertNotNull(attGroupReference.getExplicitReference());

        CdmAttributeGroupDefinition attGroupDefinition = (CdmAttributeGroupDefinition) attGroupReference.getExplicitReference();
        Assert.assertEquals(attributeGroupName, attGroupDefinition.getAttributeGroupName());

        return attGroupDefinition;
    }

    /**
     * Validates trait "has.expansionInfo.list" for array type.
     * @param attribute The type attribute.
     * @param expectedAttrName The expected attribute name.
     * @param ordinal The expected ordinal.
     * @param expansionName The expected expansion name.
     * @param memberAttribute The expected member attribute name.
     */
    public static void validateExpansionInfoTrait(CdmTypeAttributeDefinition attribute, String expectedAttrName, int ordinal, String expansionName, String memberAttribute) {
        Assert.assertEquals(expectedAttrName, attribute.getName());
        CdmTraitReference trait = (CdmTraitReference) attribute.getAppliedTraits().item("has.expansionInfo.list");
        Assert.assertNotNull(trait);
        Assert.assertEquals(trait.getArguments().fetchValue("expansionName"), expansionName);
        Assert.assertEquals(trait.getArguments().fetchValue("ordinal"), String.valueOf(ordinal));
        Assert.assertEquals(trait.getArguments().fetchValue("memberAttribute"), memberAttribute);
    }

    /**
     * Validates if the attribute context of the resolved entity matches the expected output.
     * @see ProjectionTestUtils#validateAttributeContext(List, String, String, CdmEntityDefinition, boolean)
     */
    private static void validateAttributeContext(List<String> directives, String expectedOutputPath, String entityName, CdmEntityDefinition resolvedEntity) {
        validateAttributeContext(directives, expectedOutputPath, entityName, resolvedEntity, false);
    }

    /**
     * Validates if the attribute context of the resolved entity matches the expected output.
     * @param updateExpectedOutput If true, will update the expected output txt files for all the tests that are ran.
     */
    private static void validateAttributeContext(List<String> directives, String expectedOutputPath, String entityName, CdmEntityDefinition resolvedEntity, boolean updateExpectedOutput) {
        if (resolvedEntity.getAttributeContext() == null) {
            Assert.fail("ValidateAttributeContext called with not resolved entity.");
        }

        String fileNamePrefix = "AttrCtx_" + entityName;
        Path expectedStringFilePath;
        String fileNameSuffix = getResolutionOptionNameSuffix(directives);
        String defaultFileNameSuffix = getResolutionOptionNameSuffix(new ArrayList<>());

        // Get actual text
        AttributeContextUtil attrCtxUtil = new AttributeContextUtil();
        String actualText = attrCtxUtil.getAttributeContextStrings(resolvedEntity);

        try {
            if (updateExpectedOutput) {
                expectedStringFilePath = new File(expectedOutputPath, fileNamePrefix + fileNameSuffix + ".txt").toPath();

                if (directives.size() > 0) {
                    File defaultStringFile = new File(expectedOutputPath, fileNamePrefix + defaultFileNameSuffix + ".txt");
                    String defaultText = defaultStringFile.exists() ? new String(Files.readAllBytes(defaultStringFile.toPath()), StandardCharsets.UTF_8) : null;

                    if (actualText.equals(defaultText)) {
                        final File actualFile = new File(expectedStringFilePath.toString());
                        actualFile.delete();
                    } else {
                        try (final BufferedWriter actualFileWriter = new BufferedWriter(new FileWriter(expectedStringFilePath.toFile()))) {
                            actualFileWriter.write(actualText);
                        }
                    }
                } else {
                    try (final BufferedWriter actualFileWriter = new BufferedWriter(new FileWriter(expectedStringFilePath.toFile()))) {
                        actualFileWriter.write(actualText);
                    }
                }
            } else {
                // Actual
                Path actualStringFilePath = new File(expectedOutputPath.replace("ExpectedOutput", TestHelper.getTestActualOutputFolderName()), fileNamePrefix + fileNameSuffix + ".txt").toPath();

                // Save Actual AttrCtx_*.txt and Resolved_*.cdm.json
                try (final BufferedWriter actualFileWriter = Files.newBufferedWriter(actualStringFilePath, StandardCharsets.UTF_8, StandardOpenOption.CREATE);) {
                    actualFileWriter.write(actualText);
                    actualFileWriter.flush();
                }
                resolvedEntity.getInDocument().saveAsAsync("Resolved_" + entityName + fileNameSuffix + ".cdm.json", false).join();

                // Expected
                String expectedFileNameSuffix = getResolutionOptionNameSuffix(directives);
                File expectedFile = new File(expectedOutputPath, fileNamePrefix + expectedFileNameSuffix + ".txt");

                // If a test file doesn't exist for this set of directives, fall back to the default file.
                if (!expectedFile.exists()) {
                    expectedFile = new File(expectedOutputPath, fileNamePrefix + defaultFileNameSuffix + ".txt");
                }

                final String expectedText = new String(Files.readAllBytes(expectedFile.toPath()), StandardCharsets.UTF_8);

                // Test if Actual is Equal to Expected
                Assert.assertEquals(actualText.replace("\r\n", "\n"), expectedText.replace("\r\n","\n"));
            }
        } catch (Exception e) {
            Assert.fail(e.getMessage());
        }
    }

    /**
     * Creates an entity
     */
    public static CdmEntityDefinition createEntity(CdmCorpusDefinition corpus, CdmFolderDefinition localRoot) {
        String entityName = "TestEntity";
        CdmEntityDefinition entity = corpus.makeObject(CdmObjectType.EntityDef, entityName);

        CdmDocumentDefinition entityDoc = corpus.makeObject(CdmObjectType.DocumentDef, entityName + ".cdm.json", false);
        entityDoc.getImports().add(foundationJsonPath);
        entityDoc.getDefinitions().add(entity);
        localRoot.getDocuments().add(entityDoc, entityDoc.getName());

        return entity;
    }

    /**
     * Creates a source entity for a projection
     */
    public static CdmEntityDefinition createSourceEntity(CdmCorpusDefinition corpus, CdmFolderDefinition localRoot) {
        String entityName = "SourceEntity";
        CdmEntityDefinition entity = corpus.makeObject(CdmObjectType.EntityDef, entityName);

        String attributeName1 = "id";
        CdmTypeAttributeDefinition attribute1 = corpus.makeObject(CdmObjectType.TypeAttributeDef, attributeName1);
        attribute1.setDataType(corpus.makeRef(CdmObjectType.DataTypeRef, "string", true));
        entity.getAttributes().add(attribute1);

        String attributeName2 = "name";
        CdmTypeAttributeDefinition attribute2 = corpus.makeObject(CdmObjectType.TypeAttributeDef, attributeName2);
        attribute2.setDataType(corpus.makeRef(CdmObjectType.DataTypeRef, "string", true));
        entity.getAttributes().add(attribute2);

        String attributeName3 = "value";
        CdmTypeAttributeDefinition attribute3 = corpus.makeObject(CdmObjectType.TypeAttributeDef, attributeName3);
        attribute3.setDataType(corpus.makeRef(CdmObjectType.DataTypeRef, "integer", true));
        entity.getAttributes().add(attribute3);

        String attributeName4 = "date";
        CdmTypeAttributeDefinition attribute4 = corpus.makeObject(CdmObjectType.TypeAttributeDef, attributeName4);
        attribute4.setDataType(corpus.makeRef(CdmObjectType.DataTypeRef, "date", true));
        entity.getAttributes().add(attribute4);

        CdmDocumentDefinition entityDoc = corpus.makeObject(CdmObjectType.DocumentDef, entityName + ".cdm.json", false);
        entityDoc.getImports().add(foundationJsonPath);
        entityDoc.getDefinitions().add(entity);
        localRoot.getDocuments().add(entityDoc, entityDoc.getName());

        return entity;
    }

    /**
     * Creates a projection
     */
    public static CdmProjection createProjection(CdmCorpusDefinition corpus, CdmFolderDefinition localRoot) {
        // Create an entity reference to use as the source of the projection
        CdmEntityReference projectionSource = corpus.makeObject(CdmObjectType.EntityRef, null);
        projectionSource.setExplicitReference(createSourceEntity(corpus, localRoot));

        // Create the projection
        CdmProjection projection = corpus.makeObject(CdmObjectType.ProjectionDef);
        projection.setSource(projectionSource);

        return projection;
    }
}
